google/cloud/sql/connector/InstanceConnectionManager.py

```diff
@@ -230,3 +389,72 @@ def _auth_init(self):
 
         self._credentials = scoped_credentials
         self._cloud_sql_service = cloudsql
+
+    def _perform_refresh(self) -> concurrent.futures.Future:
+        """Retrieves instance metadata and ephemeral certificate from the
+        Cloud SQL Instance.
+
+        :rtype: concurrent.future.Futures
+        :returns: A future representing the creation of an SSLcontext.
+        """
+
+        logger.debug("Entered _perform_refresh")
+
+        instance_data_task = asyncio.run_coroutine_threadsafe(
+            self._get_instance_data(), loop=self._loop
+        )
+        instance_data_task.add_done_callback(self._update_current)
+
+        return instance_data_task
+
+    async def _schedule_refresh(self, delay: int) -> concurrent.futures.Future:
+        """A coroutine that sleeps for the specified amount of time before
+        running _perform_refresh.
+
+        :type delay: int
+        :param delay: An integer representing the number of seconds for delay.
+
+        :rtype: asyncio.Task
+        :returns: A Task representing _get_instance_data.
+        """
+        logger.debug("Entering sleep")
+
+        try:
+            await asyncio.sleep(delay)
+        except asyncio.CancelledException:
+            logger.debug("Task cancelled.")
+            return None
+
+        return self._perform_refresh()
+
+    @staticmethod
+    def immediate_future(object: Any) -> concurrent.futures.Future:
+        """A static method that returns an finished future representing
+        the object passed in.
+
+        :type object: Any
+        :param object: Any object.
+
+        :rtype: concurrent.futures.Future
+        :returns: A concurrent.futures.Future representing the value passed
+            in.
+        """
+        fut: concurrent.futures.Future = concurrent.futures.Future()
+        fut.set_result(object)
+        return fut
+
+    def connect(self,) -> OpenSSL.SSL.Connection:
+        """A method that returns an OpenSSL connection to the database.
+
+        :rtype: OpenSSl.SSL.Connection
+        :returns: An OpenSSL connection to the primary IP of the database.
+        """
+        instance_data: InstanceMetadata = self._current_instance_data.result()
```

Comment:
If we are going to use a lock on `self._current` and `self._next` we need to make sure to acquire the lock every time we read/write to these attributes.

You should do something like this: 
```python
with self._lock:
    instance_data = self._current_instance_data
```

It also looks like there are some attributes being mixed up here - is it `self._current` or `self._current_instance_data`?